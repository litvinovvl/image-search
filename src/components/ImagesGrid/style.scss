.container {
  padding: 40px 80px;
  background-color: #3A71D5;
  height: 100vh;
  overflow: auto;
  transition: height 1s ease-in-out;

  &.hidden {
    height: 0;
    padding: 0;
  }

  .images-wrapper {
    display: grid;
    grid-template-columns: repeat(3, 1fr);
    grid-template-rows: 300px;
    grid-gap: 30px;
  
    & .image {
      cursor: pointer;
      width: 100%;
      height: 100%;
      object-fit: cover;
    }
  }

  .spinner-wrapper {
    height: 100%;
    display: flex;
    align-items: center;
    justify-content: center;
    color: #FFF;
    font-size: 32px;
    font-weight: bold;
  }
}

.modal {
  position: absolute;
  top: 40px;
  left: 40px;
  right: 40px;
  bottom: 40px;
  background-color: #353335;

  .modal-content {
    width: 100%;
    height: 100%;

    .image {
      width: 100%;
      height: 100%;
      object-fit: cover;
    }
  }
}

.overlay {
  position: fixed;
  top: 0;
  left: 0;
  right: 0;
  bottom: 0;
  background-color: rgba(0, 0, 0, 0.5);
}
