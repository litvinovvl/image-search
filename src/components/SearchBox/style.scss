.search-wrapper {
  background-color: #353335;
  height: 100%;
  display: flex;
  align-items: center;
  justify-content: center;

  &.shrink {
    height: 80px;
    transition: all 1s ease-in-out;
  }

  .search-input {
    height: 40px;
    width: 500px;
    border-radius: 4px;
    border: 1px solid;
    padding: 0 16px;

    &:focus-visible {
      outline: 2px solid #3A71D5;
      outline-offset: -2px;
    }

    &::placeholder {
      color: #D5D8DC;
    }
  }

  .search-btn {
    margin: 0 10px;
    height: 42px;
    padding: 12px 20px;
    background-color: #3A71D5;
    color: #FFF;
    cursor: pointer;
    border: none;
    border-radius: 4px;
    font-weight: bold;

    &:hover {
      opacity: 0.9;
    }

    &:active {
      opacity: 0.7;
    }
  }
}
