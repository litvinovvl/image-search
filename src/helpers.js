export const generateSearchPath = (query) => (
  `https://pixabay.com/api/?key=${process.env.REACT_APP_PIXABAY_API_KEY}&q=${query}&image_type=photo&pretty=true`
);

export const fetchImages = (query) => fetch(generateSearchPath(query));
