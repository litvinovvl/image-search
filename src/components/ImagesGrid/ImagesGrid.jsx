import React, { useState } from 'react';
import classNames from 'classnames';
import PropTypes from 'prop-types';
import ReactModal from 'react-modal';

import { ReactComponent as NoResultsIcon } from '../../assets/images/sad-face.svg';
import { ReactComponent as LoadingIcon } from '../../assets/images/think-face.svg';

import './style.scss';

const ImagesGrid = ({
  images, isLoading, error,
}) => {
  const [openedImage, setOpenedImage] = useState(null);
  const isModalOpened = Boolean(openedImage);

  const handleImgClick = (image) => () => setOpenedImage(image);

  const handleImgKeyPress = (image) => (e) => {
    if (e.key === 'Enter') setOpenedImage(image);
  };

  const renderImages = () => {
    if (error) {
      return (
        <div className="spinner-wrapper">
          <NoResultsIcon width={48} hanging={48} />
          Something went wrong, try again
        </div>
      );
    }

    if (isLoading) {
      return (
        <div className="spinner-wrapper">
          <LoadingIcon width={48} hanging={48} />
          Loading
        </div>
      );
    }

    if (images && images.length) {
      return (
        <div className="images-wrapper">
          {
            images.map((image) => (
              <div
                key={image.id}
                role="button"
                onClick={handleImgClick(image)}
                onKeyPress={handleImgKeyPress(image)}
                tabIndex={0}
              >
                <img
                  src={image.webformatURL}
                  alt={image.tags}
                  className="image"
                />
              </div>

            ))
          }
        </div>
      );
    }

    return (
      <div className="spinner-wrapper">
        <NoResultsIcon width={48} hanging={48} />
        No results found!
      </div>
    );
  };

  const closeModal = () => setOpenedImage(null);

  return (
    <div className={classNames('container', { hidden: !images })}>
      {renderImages()}
      <ReactModal
        className="modal"
        overlayClassName="overlay"
        isOpen={isModalOpened}
        onRequestClose={closeModal}
      >
        {isModalOpened && (
          <div className="modal-content">
            <img src={openedImage.largeImageURL} alt={openedImage.tags} className="image" />
          </div>
        )}
      </ReactModal>
    </div>
  );
};

ImagesGrid.defaultProps = {
  images: null,
  isLoading: false,
  error: false,
};

ImagesGrid.propTypes = {
  images: PropTypes.arrayOf(PropTypes.object),
  isLoading: PropTypes.bool,
  error: PropTypes.bool,
};

export default ImagesGrid;
