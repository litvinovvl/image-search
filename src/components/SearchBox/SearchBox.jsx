import React, { useState } from 'react';
import classnames from 'classnames';
import PropTypes from 'prop-types';

import './style.scss';

const SearchBox = ({ isLoading, fullScreen, onSubmit }) => {
  const [query, setQuery] = useState('');

  const handleSubmit = (e) => {
    e.preventDefault();
    onSubmit(query);
  };

  const handleChange = ({ target: { value } }) => setQuery(value);

  return (
    <form onSubmit={handleSubmit} className={classnames('search-wrapper', { shrink: !fullScreen })}>
      <input
        className="search-input"
        placeholder="What images would ypu like to see on Pixabay?"
        onChange={handleChange}
      />
      <button type="submit" className="search-btn">
        {isLoading ? 'Searching' : 'Search'}
      </button>
    </form>
  );
};

SearchBox.defaultProps = {
  fullScreen: true,
  isLoading: false,
};

SearchBox.propTypes = {
  fullScreen: PropTypes.bool,
  isLoading: PropTypes.bool,
  onSubmit: PropTypes.func.isRequired,
};

export default SearchBox;
