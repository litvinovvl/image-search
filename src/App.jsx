import React, { useState } from 'react';

import SearchBox from './components/SearchBox';
import ImagesGrid from './components/ImagesGrid';
import { fetchImages } from './helpers';

import './styles.scss';

const App = () => {
  const [lastQuery, setLastQuery] = useState('');
  const [isLoading, setIsLoading] = useState(false);
  const [error, setError] = useState(false);
  const [images, setImages] = useState(null);
  const shouldRenderImages = Boolean(images);

  const searchImages = async (query) => {
    if (lastQuery === query) return;

    setError(false);
    setIsLoading(true);

    try {
      const response = await fetchImages(query);
      const data = await response.json();
      setLastQuery(query);
      setImages(data.hits);
      setIsLoading(false);
    } catch (err) {
      setImages(images || []);
      setError(true);
      setIsLoading(false);
    }
  };

  return (
    <div className="wrapper">
      <SearchBox
        fullScreen={!shouldRenderImages}
        isLoading={isLoading}
        onSubmit={searchImages}
      />
      <ImagesGrid images={images} isLoading={isLoading} error={error} />
    </div>
  );
};

export default App;
