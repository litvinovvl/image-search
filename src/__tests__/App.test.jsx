import React from 'react';
import { shallow } from 'enzyme';

import App from '../App';
import ImagesGrid from '../components/ImagesGrid';
import SearchBox from '../components/SearchBox';
import * as helpers from '../helpers';

const data = { hits: [{ id: 1, webformatURL: 'test_url', tags: 'tags' }] };

jest.mock('../helpers', () => ({
  fetchImages: jest.fn(() => Promise.resolve({
    json: () => Promise.resolve({ hits: [{ id: 1, webformatURL: 'test_url', tags: 'tags' }] }),
  })),
}));

describe('App', () => {
  beforeEach(jest.clearAllMocks);

  it('should render App properly', () => {
    const component = shallow(<App />);

    expect(component).toMatchSnapshot();
  });

  it('should pass right props to inner components initially', () => {
    const component = shallow(<App />);
    const Search = component.find(SearchBox);
    const Grid = component.find(ImagesGrid);

    expect(Search.prop('fullScreen')).toBeTruthy();
    expect(Search.prop('onSubmit')).toEqual(expect.any(Function));

    expect(Grid.prop('images')).toBe(null);
    expect(Grid.prop('isLoading')).toBeFalsy();
    expect(Grid.prop('error')).toBeFalsy();
  });

  it('should pass right props to inner components after images loaded', async () => {
    const component = shallow(<App />);
    const Search = component.find(SearchBox);

    Search.props().onSubmit('query');
    component.update();

    const PendingGrid = component.find(ImagesGrid);
    expect(PendingGrid.prop('isLoading')).toBeTruthy();

    await Search.props().onSubmit('query');
    component.update();

    const UpdatedSearch = component.find(SearchBox);
    const UpdatedGrid = component.find(ImagesGrid);

    expect(UpdatedSearch.prop('fullScreen')).toBeFalsy();

    expect(UpdatedGrid.prop('images')).toEqual(data.hits);
    expect(UpdatedGrid.prop('isLoading')).toBeFalsy();
    expect(UpdatedGrid.prop('error')).toBeFalsy();
  });

  it('should not refetch if query was not changed', async () => {
    const component = shallow(<App />);

    await component.find(SearchBox).prop('onSubmit')('query');
    helpers.fetchImages.mockClear();
    await component.find(SearchBox).prop('onSubmit')('query');

    expect(helpers.fetchImages).not.toHaveBeenCalled();
  });

  it('should pass error to inner component after images load failed', async () => {
    helpers.fetchImages.mockImplementation(jest.fn(Promise.reject));

    const component = shallow(<App />);
    const Search = component.find(SearchBox);

    await Search.props().onSubmit('query');
    component.update();

    const UpdatedGrid = component.find(ImagesGrid);
    expect(UpdatedGrid.prop('isLoading')).toBeFalsy();
    expect(UpdatedGrid.prop('error')).toBeTruthy();
  });
});
