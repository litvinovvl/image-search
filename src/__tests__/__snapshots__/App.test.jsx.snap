// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App should render App properly 1`] = `
<div
  className="wrapper"
>
  <SearchBox
    fullScreen={true}
    isLoading={false}
    onSubmit={[Function]}
  />
  <ImagesGrid
    error={false}
    images={null}
    isLoading={false}
  />
</div>
`;
